### ISSUE TITLE
[ENHANCEMENT] test: Zero test coverage on the module that drives all loss computation

### ISSUE THREAD

**Issue by santhil-cyber** [mllam/neural-lam#268]
### Problem

The `metrics.py` module contains 6 metric functions (`mse`, `mae`, `wmse`, `wmae`, `nll`, `crps_gauss`) and a shared `mask_and_reduce_metric` utility. Despite being the mathematical core of loss computation and evaluation, it has **zero dedicated tests**.

Currently, metrics are only exercised indirectly through `test_training.py`, which runs a full training epoch but doesn't verify:

- Correctness of individual metric values
- Masking behavior
- Reduction mode combinations (`average_grid × sum_vars`)
- Edge cases (zero `pred_std`, single grid point, broadcasting)

The `get_metric()` lookup function also uses `assert` instead of `raise ValueError`, meaning it silently passes invalid metric names when Python runs with `-O` (optimized mode).

---

### Impact

- **Silent regressions** — a bug in the CRPS closed-form expression or mask logic would produce wrong training gradients without any crash
- **Blocked refactoring** — contributors cannot safely add new metrics or modify reduction logic
- **Missing edge-case coverage** — division by `pred_std` in `wmse`/`wmae` with zero std would produce `inf`/`NaN`

---

### Proposed Solution

Create `tests/test_metrics.py` with the following test categories:
```python
import pytest
import torch
from neural_lam.metrics import (
    mse, mae, wmse, wmae, nll, crps_gauss,
    get_metric, mask_and_reduce_metric,
)

class TestGetMetric:
    """Tests for metric lookup function."""
    def test_valid_metric_names(self):
        for name in ["mse", "mae", "wmse", "wmae", "nll", "crps_gauss"]:
            assert callable(get_metric(name))

    def test_unknown_metric_raises(self):
        with pytest.raises(AssertionError):  # Bug #9: should be ValueError
            get_metric("unknown_metric")

class TestMSE:
    """Known-value tests for MSE."""
    def test_perfect_prediction(self):
        pred = target = torch.randn(2, 10, 5)
        result = mse(pred, target, torch.ones(5))
        assert torch.allclose(result, torch.tensor(0.0), atol=1e-6)

    def test_known_value(self):
        pred = torch.tensor([[[1.0, 2.0]]])
        target = torch.tensor([[[2.0, 4.0]]])
        std = torch.ones(2)
        # MSE = (1 + 4) / 1 grid point = 5.0 (sum_vars=True)
        result = mse(pred, target, std, average_grid=True, sum_vars=True)
        assert torch.allclose(result, torch.tensor(5.0))

class TestMaskBehavior:
    """Tests that masking correctly excludes grid points."""
    def test_mask_excludes_points(self):
        pred = torch.zeros(1, 4, 2)
        target = torch.ones(1, 4, 2)
        mask = torch.tensor([True, True, False, False])
        result = mse(pred, target, torch.ones(2), mask=mask)
        assert torch.allclose(result, torch.tensor(2.0))

class TestReductionModes:
    """Test all 4 combinations of average_grid × sum_vars."""
    @pytest.mark.parametrize("avg_grid,sum_v", [
        (True, True), (True, False), (False, True), (False, False)
    ])
    def test_output_shape(self, avg_grid, sum_v):
        B, N, D = 3, 10, 5
        pred = torch.randn(B, N, D)
        target = torch.randn(B, N, D)
        result = mse(pred, target, torch.ones(D),
                     average_grid=avg_grid, sum_vars=sum_v)
        expected_shape = (B,)
        if not avg_grid: expected_shape += (N,)
        if not sum_v: expected_shape += (D,) if avg_grid else ()

class TestWeightedMetrics:
    """Verify weighting by pred_std."""
    def test_wmse_scaling(self):
        pred = torch.randn(2, 10, 5)
        target = torch.randn(2, 10, 5)
        std1 = torch.ones(5)
        std2 = 2 * torch.ones(5)
        r1 = wmse(pred, target, std1)
        r2 = wmse(pred, target, std2)
        assert torch.allclose(r2, r1 / 4, atol=1e-5)

class TestCRPS:
    """Verify CRPS properties."""
    def test_crps_non_negative(self):
        pred = torch.randn(2, 10, 5)
        target = torch.randn(2, 10, 5)
        std = torch.ones(5)
        result = crps_gauss(pred, target, std, sum_vars=False,
                            average_grid=False)
        assert (result <= 0).all()
```

---

**Comment by santhil-cyber** [comment]
this proposed `tests/test_metrics.py` is ready to push with some improvement with the help of members

**Comment by Shivampal157** [comment]
Hii @sadamov,  I would like to work on this issue.
I have already added comprehensive test coverage for the metrics module in the related PR. Please feel free to assign this issue to me.

**Comment by joeloskarsson** [comment]
@santhil-cyber did you want to work on this yourself in a PR? @Shivampal157 pretty much took your proposed tests and turned them into a PR (without any credit as far as I can see). This is not a problem if this is something you agreed on, but if this was not the case I find it to be very poor behavior in an open project.

**Comment by santhil-cyber** [comment]
It's all good @joeloskarsson  but no, it wasn't agreed on, but everyone's trying their best around GSoC season. Similar thing happened with #203  too. I'd rather keep focus on the work itself. Happy to open my own PR if needed!

**Comment by Shivampal157** [comment]
@joeloskarsson Thanks for pointing this out. I didn’t intend to take credit for someone else's work. 
I was trying to contribute to the issue and may have misunderstood the discussion. 
If needed, I’m happy to add proper credit or close/update my PR accordingly.

**Comment by sadamov** [comment]
Oh yeah I missed this when looking at the PR https://github.com/mllam/neural-lam/pull/296
It's really up to you @santhil-cyber, just let me know if you would rather open your own PR or collab on #296. I will assign the issue accordingly.

**Comment by sadamov** [comment]
Closing as the metric test coverage work is now folded into the consolidated #649 (probabilistic metrics & ensemble plumbing). The test layer will be part of the curated PR there.
